src/services/upower/device.hpp

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <qcontainerfwd.h>
+#include <qobject.h>
+#include <qqmlintegration.h>
+#include <qtmetamacros.h>
+#include <qtypes.h>
+
+#include "../../core/doc.hpp"
+#include "../../dbus/properties.hpp"
+#include "dbus_device.h"
+
+namespace qs::service::upower {
+
+class UPowerDeviceState: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		Charging = 1,
+		Discharging = 2,
+		Empty = 3,
+		FullyCharged = 4,
+		PendingCharge = 5,
+		PendingDischarge = 6,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceState::Enum status);
+};
+
+class UPowerDeviceType: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		LinePower = 1,
+		Battery = 2,
+		Ups = 3,
+		Monitor = 4,
+		Mouse = 5,
+		Keyboard = 6,
+		Pda = 7,
+		Phone = 8,
+		MediaPlayer = 9,
+		Tablet = 10,
+		Computer = 11,
+		GamingInput = 12,
+		Pen = 13,
+		Touchpad = 14,
+		Modem = 15,
+		Network = 16,
+		Headset = 17,
+		Speakers = 18,
+		Headphones = 19,
+		Video = 20,
+		OtherAudio = 21,
+		RemoteControl = 22,
+		Printer = 23,
+		Scanner = 24,
+		Camera = 25,
+		Wearable = 26,
+		Toy = 27,
+		BluetoothGeneric = 28,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceType::Enum type);
+};
+
+///! A device exposed through the UPower system service.
+class UPowerDevice: public QObject {
+	Q_OBJECT;
+	// clang-format off
+	/// The type of power device.
+	///
+	/// > [!INFO] When checking if the device is a battery, also verify that `powerSupply` is set to `true`.
+	Q_PROPERTY(UPowerDeviceType::Enum type READ type NOTIFY typeChanged);
+	/// If the device is a power supply.
+	Q_PROPERTY(bool powerSupply READ powerSupply NOTIFY powerSupplyChanged);
```

Comment:
Is this redundant with `type` or does it span more than just `LinePower`?

Reply:
According to the DBus API reference, it's necessary to make sure both `powerSupply` is true and `type` is `Battery` for it to be considered a valid device. To avoid confusion, there are two ways I can think of making this simpler: 
1. Replacing it with an `isBattery` property:
   * This property will return true if both requirements are met (`powerSupply` is true and `type` is `Battery`).
   * This approach may introduce confusion if it coexists with the `type` property. To mitigate this, we might consider removing the `type` property, assuming the service is primarily used for batteries.
2. Masking it in the `type` property:
   * Modify the `type` property to return `Battery` only if `powerSupply` is also true. Otherwise, it will return `Unknown`.
   * This approach simplifies the interface by consolidating the logic into the `type` property, but introduces some hidden behavior.

I'm leaning more towards the first approach and removing the `type` property as it simplifies the interface while focusing more on battery devices, but I am very much open to any other ideas.

Reply:
Yeah thats a bit weird. maybe make an `isLaptopBattery` property that just checks both. I'd leave the `type` property though.

Reply:
Checked my power adapter with dbus and it seems to confirm this. `powerSupply` means "is supplying power to this computer"

Reply:
Seems good!